block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagRepresentationSyntax.scala

```diff
@@ -19,9 +19,17 @@ final class BlockDagRepresentationOps[F[_]: Sync](
     // BlockDagRepresentation extensions / syntax
     private val dag: BlockDagRepresentation[F]
 ) {
-  // Get block metadata, "unsafe" because method expects block already in the DAG.
-  def lookupUnsafe(hash: BlockHash): F[BlockMetadata] = {
-    def errMsg = s"DAG storage is missing hash ${PrettyPrinter.buildString(hash)}"
+
+  /**
+    * Get block metadata, "unsafe" because method expects block already in the DAG.
+    *
+    * Unfortunately there is no way to get stack trace when error is thrown in async execution.
+    * Monix does not have support and cats.effect support is in creation.
+    * https://github.com/typelevel/cats-effect/pull/854
+    * So extra source parameter is a desperate measure to indicate who is the caller.
+    */
+  def lookupUnsafe(hash: BlockHash)(implicit source: sourcecode.Enclosing): F[BlockMetadata] = {
```

Comment:
Is this implicit something always available or we might need to provide it in some cases?

Reply:
Yes, this is filled in automatically or it can be manually supplied.

> All these are available both via () and as implicits, e.g. sourcecode.File can be summoned via sourcecode.File() or implicitly[sourcecode.File].value. This also means you can define functions that pull in this information automatically
https://github.com/lihaoyi/sourcecode#overview

We can print the source file itself with line number. Maybe this would be better? Although it will be longer string so maybe it can be in the next line?
https://github.com/lihaoyi/sourcecode#logging

I'll also try something similar like in this example. To override monix Task implementation with addition of trace. It should be without runtime overhead.
https://www.schibsted.pl/blog/tracing-back-scala-future-chains/

Reply:
Does this relates to the second proposal? https://github.com/rchain/rchain/blob/5a7f3c67d80811be2f5d028e4dab2a02bfa8ace0/node/src/main/scala/coop/rchain/node/NodeRuntime.scala#L96-L99 

I believe we have extended Task for tracing spans

Reply:
Also this might be a good addon <https://kamon.io/docs/v1/instrumentation/logback/> we can have the same spans info that we have in Jaeger - in logs

Reply:
Yes, this is similar situation with metrics trace which must carry the state through threads. I'm wondering if we can replace our custom solution with some existing library.